bundles/org.eclipse.equinox.common/src/org/eclipse/core/runtime/URIUtil.java

```diff
@@ -285,18 +299,105 @@ public static URI toURI(URL url) throws URISyntaxException {
 			//ensure there is a leading slash to handle common malformed URLs such as file:c:/tmp
 			if (pathString.indexOf('/') != 0)
 				pathString = '/' + pathString;
-			else if (pathString.startsWith(UNC_PREFIX) && !pathString.startsWith(UNC_PREFIX, 2)) {
-				//URL encodes UNC path with two slashes, but URI uses four (see bug 207103)
-				pathString = ensureUNCPath(pathString);
-			}
-			return new URI(SCHEME_FILE, null, pathString, null);
+			return toURI(SCHEME_FILE, null, pathString, null);
 		}
 		try {
 			return new URI(url.toExternalForm());
 		} catch (URISyntaxException e) {
 			//try multi-argument URI constructor to perform encoding
-			return new URI(url.getProtocol(), url.getUserInfo(), url.getHost(), url.getPort(), url.getPath(), url.getQuery(), url.getRef());
+			return toURI(url.getProtocol(), url.getUserInfo(), url.getHost(), url.getPort(), url.getPath(),
+					url.getQuery(), url.getRef());
+		}
+	}
+
+	/**
+	 * An UNC-safe factory the the URI-ctor
+	 * 
+	 * @param scheme   Scheme name
+	 * @param ssp      Scheme-specific part
+	 * @param fragment Fragment
+	 * @return The new URI from the given components
+	 * @throws URISyntaxException If the URI string constructed from the given
+	 *                            components violates RFC 2396
+	 * @see java.net.URI#URI(String, String, String)
+	 * @since 3.17
+	 */
+	public static URI toURI(String scheme, String ssp, String fragment) throws URISyntaxException {
```

Comment:
Is there any reason to have **3 new API** methods with lot of String arguments that could be wrongly used? Would it be enough to have one, or better leave all of them **private** to avoid confusion for API clients?

Reply:
The motivation is, that these are drop-in replacements for the original constructors which have lots of issues with UNC-Paths - thus they must have the same interface.

Reply:
Would it help to document that more thoroughly? I can also add tests that expect that exception. :)

Reply:
As it seems mostly to be a fix, and that UNC is stanard enough, can't we just improve the existing methods to work as expected instead of introducing new ones?

Reply:
@mickaelistria  I'd love to. But this is a java-lib issue since forever (I'd have to check, but I think one of the bugs I found was filed against Java 5.0). All the tickets are wontfix, because it might break things. I've found multiple local workarounds in the CDT and Eclipse Codebase and it would be nice to have this at a central place.

Reply:
It's still not clear to me why new public method entry points in URIUtil are necessary. Can the existing ones in URIUtil be fixed to work better with UNC?

Reply:
I think we have a misunderstanding here. This is not about improving the existing implementation (which already has a spacial handling to convert URLs into URIs), but extending it, so it can also be used as drop-in replacement for the broken [java.net.URI-ctors](https://docs.oracle.com/javase/7/docs/api/java/net/URI.html#constructor_summary).

Reply:
Looking at the referenced Bug descriptors, I can't see that anything is broken with URI constructor per se and it seems the only "value" is that you add more slashes. So why are these slashes missing in the first place? The referenced bugs seem to indicate inappropriate use of URI.normalize() and suggest using Path#normalize instead...

Reply:
I improved the commit message and the tests. I hope things are clearer now.